test/run

```diff
@@ -336,153 +402,147 @@ run_s2i_build_express_webapp() {
+    fi
+    printf -v test_short_summary "%s %s for '%s' %s\n" "${test_short_summary}" "${test_msg}" "${APP_NAME}" "$test_case"
+    [ -n "${FAIL_QUICKLY:-}" ] && cleanup "${APP_NAME}" && return 1
+  done;
+}
 
-# Test that the development dependencies (nodemon) have NOT been removed
-devdep=$(docker run --rm ${IMAGE_NAME}-testapp /bin/bash -c "test -d ~/node_modules/nodemon")
+# Build the application image twice to ensure the 'save-artifacts' and
+# 'restore-artifacts' scripts are working properly
+prepare app
+echo "Testing the production image build"
+run_s2i_build
 check_result $?
 
-# Test that the npm cache has NOT been cleared
-devdep=$(docker run --rm ${IMAGE_NAME}-testapp /bin/bash -c "test -d \$(npm config get cache)")
-check_result $?
+TEST_SET=${TESTS:-$TEST_LIST_APP} run_all_tests "app"
 
-kill_test_application
+echo "Testing the development image build: s2i build -e \"NODE_ENV=development\")"
+run_s2i_build "-e NODE_ENV=development"
+check_result $?
 
-test_dev_mode app true development
-test_dev_mode app false development
+TEST_SET=${TESTS:-$TEST_LIST_DEV} run_all_tests "development"
 
 echo "Testing the development image build: s2i build -e \"DEV_MODE=true\")"
 run_s2i_build "-e DEV_MODE=true"
 check_result $?
 
-# Verify that the HTTP connection can be established to test application container
-run_test_application app
-wait_for_cid
-
-test_connection
-check_result $?
-
-# Test that the development dependencies (nodemon) have NOT been removed
-devdep=$(docker run --rm ${IMAGE_NAME}-testapp /bin/bash -c "test -d ~/node_modules/nodemon")
-check_result $?
-
-# Test that the npm cache has NOT been cleared
-devdep=$(docker run --rm ${IMAGE_NAME}-testapp /bin/bash -c "test -d \$(npm config get cache)")
-check_result $?
-
-kill_test_application
-
-test_dev_mode app true development
-test_dev_mode app false production
-
 echo "Testing proxy safe logging..."
 prepare hw
 run_s2i_build_proxy http://user.password@0.0.0.0:8000 https://user.password@0.0.0.0:8000 > /tmp/build-log 2>&1
-if [[ $(grep redacted /tmp/build-log | wc -l) -eq 4 ]]; then
-    grep redacted /tmp/build-log
-    check_result 0
-else
-    echo "Some proxy log-in credentials were left in log file"
-    grep Setting /tmp/build-log
-    check_result 1
-fi
-
-run_test_application hw
-wait_for_cid
-kill_test_application
-
-echo "Testing incremental build"
-test_incremental_build
-
-echo "Running express webapp test"
-run_s2i_build_express_webapp ; check_result $?
 
-echo "Testing npm availability"
-ct_npm_works
-check_result $?
-
-echo "Running express.js test suite"
-prepare express
-run_s2i_build_express; check_result $?
-run_express_test_suite; check_result $?
-
-# These tests only make sense for rhscl
-if [ "$OS" == "rhel7" ] || [ "$OS" == "centos7" ]; then
-  # autocleanup only enabled here as only the following tests so far use it
-  CID_FILE_DIR=$(mktemp -d)
-  ct_enable_cleanup
-
-  echo "Testing npm availability in Dockerfile"
-  ct_binary_found_from_df npm
```

Comment:
ct_binary_found_from_df is no more useful?

Reply:
Sorry I have missed that;) I will fix.